### PR TITLE
KAFKA-4010; ConfigDef.toRst() to have grouped sections with dependents info

### PR DESCRIPTION
**Pull request by rekhajoshm**
- Added sort method with group order
- Added dependents info

### CODE REVIEW

**Comment by shikhar** [comment]
Thanks @rekhajoshm for the patch! Reviewing now.

Not sure where all the stray merge commits are from, would you be able to clean that up?

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:1054]: The order in each group should be based on... `ConfigKey.orderInGroup` :-)

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: this doesn't seem right -- the JIRA suggests that the group should be a section header for all config keys with that group

**Comment by shikhar** [comment]
What I had in mind for the output is, if there are config keys `key.a`, `key.b`, `key.c`, `key.d`, with `key.a`, `key.b`, `key.c` in group `Group X` and `key.d` in group `Group Y`, the output being like

```
Group X
^^^^^^^^

``key.a``
  ...

``key.b``
  ...

``key.c``
  ...


Group Y
^^^^^^^^

``key.d``
  ...

```

so that the doc has section headers for each group. It would also be nice if for each group we are able to,

If you want an example to play with, https://gist.github.com/shikhar/3fd0fdc410c18dca7524f81ac242a266 can be pasted into `org.apache.kafka.common.config` and has a `main()` that calls `toRst()`. I'd expect to be able to generate the contents of https://raw.githubusercontent.com/confluentinc/kafka-connect-jdbc/master/docs/sink_config_options.rst from that.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: I think this should be a single line break

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: The dependent key name should be in backticks. And perhaps comma-delimited rather than tabs.

**Comment by shikhar** [comment]
I think we'd probably want to preserve a version that does not exercise any of the newer metadata (`group`, `orderInGroup`, `dependents`) since those are optional. Perhaps this version should be a new `toEnrichedRst()` call. /cc @Ishiihara @ewencp

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Also we should omit the Dependents item from the list if they are empty.

**Comment by rekhajoshm** [comment]
Thanks @shikhar for your review comments.Updated. Please have a look. thanks

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: need a newline after the group before the underline, and a double newline after the underline

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: these 2 lines shouldn't be here

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: There's a trailing comma issue

**Comment by shikhar** [comment]
Thanks for your iterating fast @rekhajoshm, left a few more comments. Other than those, this is shaping up.

**Comment by shikhar** [comment]
Before any merge I think we'd want to sort out the extra merge commits that crept into this branch. Not sure what caused that. To fix I'd probably do something like -
- checkout trunk, get it current with apache/kafka repo, resets as required
- checkout this branch, note sha of HEAD
- git reset --hard trunk
- git cherry-pick $sha_from_above
- git push -f

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Currently this is reverse order. Probably better to just do `return Integer.compare(k1.orderInGroup, k2.orderInGroup)` here.

**Comment by rekhajoshm** [comment]
Thank you @shikhar for your review comments.As my git history had multiple unmerged commits, reset/cherry pick did not work smoothly..I did a reset to commit id one prior to my first unmerged commit(of months back) to get just this commit.Please have a look.Mostly committers themselves cherry pick my last commit :-) Thanks.

**Comment by shikhar** [comment]
@rekhajoshm ah okay, I wasn't familiar with the committers' workflow :-)

**Comment by shikhar** [comment]
LGTM!

/cc @ewencp

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: At this point we have a lot of repeated code with `toRst()`. Would probably make sense to refactor to share the majority of the code for each config's main set of fields.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: `group` could be `null`

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Need to check if `group` is `null` in both `k1` and `k2`. Using this on, e.g., `DistributedConfig` from Kafka Connect doesn't currently work.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: I figured if you're calling `toEnrichedRst()` the new 0.10 fields are expected to be set

Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Groups in particular may not make sense in some cases. Some connectors have only a handful of options so grouping them isn't particularly useful.

**Comment by rekhajoshm** [comment]
Thank you @ewencp @shikhar for your review comments.sorry, had lost thread of this, updated.Please have a look.thanks.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: perhaps a better name is `getConfigKeyRst()`

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: The second newline should be left for the caller, as it otherwise causes an extra line before 'Dependents' in the enriched RST

**Review by shikhar — CHANGES_REQUESTED** [review]

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Also we can just pass in the `StringBuilder` as an argument rather than create a new one here

**Comment by rekhajoshm** [comment]
@shikhar updated, please have a look.thanks

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (shikhar) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:930]: Since now the `StringBuilder` is being appended to directly, the `b.append` here would cause duplication of the accumulated content I think

**Comment by shikhar** [comment]
@rekhajoshm can you please add a test as well for each of `toRst` and `toEnrichedRst`, exercising the different configuration options and how they should be formatted? Manually verifying the output is messy, so we should have a test.

**Comment by ewencp** [comment]
@rekhajoshm I've merged https://github.com/apache/kafka/pull/1964 as a slightly updated version of this with you still credited. Would you mind closing this version of the PR out?

**Comment by hachikuji** [comment]
Closing since this was already merged in #1964.

**Pull request closed without merging**
